### PR TITLE
Refactored component hierarchies and initialization

### PR DESCRIPTION
**Pull request by agronholm**

### CODE REVIEW

File: docs/userguide/components.rst [hunk 1]
```diff
@@ -28,19 +28,17 @@ In order to speed up the startup process and to prevent any deadlocks, component
...
-however, that if that resource is never added by any component in the context, the
-application start-up will time out.
+Component hierarchies
+---------------------
+
+Any Asphalt component can have *child components* added to it. When a component is
+started by :func:`start_component`, its child components are started first, and only
+then is the parent component itself started. The idea is that child components provide
+services that either the parent component, or another child component, require to
+provide their own services and/or resources call is launched in its own task. Therefore
```
Comment (davidbrochart) [inline docs/userguide/components.rst:38]: I don't understand this sentence:
"The idea is that child components provide services that either the parent component, or another child component, require to provide their own services and/or resources call is launched in its own task".

Comment (agronholm) [reply docs/userguide/components.rst:38]: I think I should explain the startup sequence differently in this document.

Comment (agronholm) [reply docs/userguide/components.rst]: I mostly rewrote the docs for better clarity. I hope that addresses the issue.

File: docs/userguide/deployment.rst [hunk 2]
```diff
@@ -49,29 +49,29 @@ A production-ready configuration file should contain at least the following opti
...
-      data_directory: /some/file/somewhere
+      config:
+        data_directory: /some/file/somewhere
       components:
         mailer:
-          host: smtp.mycompany.com
-          ssl: true
+          config:
+            host: smtp.mycompany.com
+            ssl: true
```
Comment (davidbrochart) [inline docs/userguide/deployment.rst:71]: Are you sure about these changes? It seems it still works the old way:
```
        mailer:
          host: smtp.mycompany.com
          ssl: true
```

Comment (agronholm) [reply docs/userguide/deployment.rst:71]: This was something I was planning but I never implemented it. Good catch.

**Review by davidbrochart — APPROVED** [review]

**Comment by agronholm** [comment]
Thanks for the review!

**Pull request merged**